include/irr/video/IGPUObjectFromAssetConverter.h

```diff
@@ -408,6 +413,152 @@ auto IGPUObjectFromAssetConverter::create(asset::ICPUSpecializedShader** const _
     return res;
 }
 
+auto IGPUObjectFromAssetConverter::create(asset::ICPUBufferView** const _begin, asset::ICPUBufferView** const _end) -> created_gpu_object_array<asset::ICPUBufferView>
+{
+    const auto assetCount = std::distance(_begin, _end);
+    auto res = core::make_refctd_dynamic_array<created_gpu_object_array<asset::ICPUBufferView> >(assetCount);
+
+    core::vector<asset::ICPUBuffer*> cpuBufs(assetCount, nullptr);
+    for (ptrdiff_t i = 0u; i < assetCount; ++i)
+        cpuBufs[i] = _begin[i]->getUnderlyingBuffer();
+
+    core::vector<size_t> redirs = eliminateDuplicatesAndGenRedirs(cpuBufs);
+    auto gpuBufs = getGPUObjectsFromAssets(cpuBufs.data(), cpuBufs.data()+cpuBufs.size());
+
+    for (ptrdiff_t i = 0u; i < assetCount; ++i)
+    {
+        asset::ICPUBufferView* cpubufview = _begin[i];
+        IGPUOffsetBufferPair* gpubuf = (*gpuBufs)[redirs[i]].get();
+        (*res)[i] = m_driver->createGPUBufferView(gpubuf->getBuffer(), cpubufview->getFormat(), gpubuf->getOffset() + cpubufview->getOffsetInBuffer(), cpubufview->getByteSize());;
+    }
+
+    return res;
+}
+
+auto IGPUObjectFromAssetConverter::create(asset::ICPUDescriptorSetLayout** const _begin, asset::ICPUDescriptorSetLayout** const _end) -> created_gpu_object_array<asset::ICPUDescriptorSetLayout>
+{
+    const auto assetCount = std::distance(_begin, _end);
+    auto res = core::make_refctd_dynamic_array<created_gpu_object_array<asset::ICPUDescriptorSet> >(assetCount);
+
+    core::vector<asset::ICPUSampler*> cpuSamplers;//immutable samplers
+    size_t maxSamplers = 0ull;
+    size_t maxBindingsPerDescSet = 0ull;
+    size_t maxSamplersPerDescSet = 0u;
+    for (auto dsl : core::SRange<asset::ICPUDescriptorSetLayout*>(_begin, _end))
+    {
+        size_t samplersInDS = 0u;
+        for (const auto& bnd : dsl->getBindings()) {
+            const uint32_t samplerCnt = bnd.samplers ? bnd.count : 0u;
+            maxSamplers += samplerCnt;
+            samplersInDS = samplerCnt;
+        }
+        maxBindingsPerDescSet = std::max(maxBindingsPerDescSet, dsl->getBindings().length());
+        maxSamplersPerDescSet = std::max(maxSamplersPerDescSet, samplersInDS);
+    }
+    cpuSamplers.reserve(maxSamplers);
+
+    for (auto dsl : core::SRange<asset::ICPUDescriptorSetLayout*>(_begin, _end))
+    {
+        for (const auto& bnd : dsl->getBindings())
+        {
+            if (bnd.samplers)
+            {
+                for (uint32_t i = 0u; i < bnd.count; ++i)
+                    cpuSamplers.push_back(bnd.samplers[i].get());
+            }
+        }
+    }
+
+    core::vector<size_t> redirs = eliminateDuplicatesAndGenRedirs(cpuSamplers);
+    auto gpuSamplers = getGPUObjectsFromAssets(cpuSamplers.data(), cpuSamplers.data()+cpuSamplers.size());
+    size_t gpuSmplrIter = 0ull;
+
+    using gpu_bindings_array_t = core::smart_refctd_dynamic_array<IGPUDescriptorSetLayout::SBinding>;
```

Comment:
what's so different about `IGPUDescriptorSetLayout::SBinding` from `ICPUDescriptorSetLayout::SBinding` ???

Reply:
GPUSampler vs CPUSampler

Reply:
so that whole expensive loop is just because the immutable CPU sampler needs to be swapped for a GPU sampler?

Reply:
probably yes, I cant see other reason now